source/host_resolver.c

```diff
@@ -1473,8 +1475,9 @@ static int default_resolve_host(
         if (aws_array_list_length(&callback_address_list)) {
             res(resolver, host_name, AWS_OP_SUCCESS, &callback_address_list, user_data);
         } else {
-            res(resolver, host_name, aws_last_error(), NULL, user_data);
-            result = AWS_OP_ERR;
+            int last_error = aws_last_error();
+            int error_code = (last_error != AWS_ERROR_SUCCESS) ? last_error : AWS_IO_DNS_QUERY_FAILED;
```

Comment:
this is not a good way to check whether aws_last_error has been set

nowhere in our codebase do we ever bother to reset aws_last_error back to 0/SUCCESS. When a thread starts, it's 0, but after that it just holds whatever error code was last set, even if it's an old stale value from some other function called long before the function we're currently in

Whenever we call `aws_last_error()`, we need to be sure that something in our current callstack set the error. Usually we ensure that by immediately doing `goto error` after a function reports an error. But this code seems to do a whole bunch of stuff ... and at the end if things aren't perfect it's just kind of hoping (???) that an error was set along the way.

Reply:
Thanks, that's really informative.